services/src/datasets/external/netcdfcf/mod.rs

```diff
@@ -840,103 +855,270 @@ impl DataProvider for NetCdfCfDataProvider {
     }
 }
 
+#[derive(Deserialize, Serialize, Debug, Clone)]
+enum NetCdfLayerCollectionId {
+    Path {
+        path: PathBuf,
+    },
+    Group {
+        path: PathBuf,
+        groups: Vec<String>,
+    },
+    Entity {
+        path: PathBuf,
+        groups: Vec<String>,
+        entity: usize,
+    },
+}
+
+impl TryFrom<NetCdfLayerCollectionId> for LayerCollectionId {
+    type Error = crate::error::Error;
+
+    fn try_from(id: NetCdfLayerCollectionId) -> crate::error::Result<Self> {
+        Ok(LayerCollectionId(serde_json::to_string(&id)?))
+    }
+}
+
+impl TryFrom<NetCdfLayerCollectionId> for LayerId {
+    type Error = crate::error::Error;
+
+    fn try_from(id: NetCdfLayerCollectionId) -> crate::error::Result<Self> {
+        Ok(LayerId(serde_json::to_string(&id)?))
+    }
+}
+
+async fn listing_from_dir(base: &Path, path: &Path) -> crate::error::Result<Vec<CollectionItem>> {
+    let mut dir = tokio::fs::read_dir(base.join(&path)).await?;
+
+    let mut items = vec![];
+    while let Some(entry) = dir.next_entry().await? {
+        if entry.path().is_dir() || entry.path().extension() == Some("nc".as_ref()) {
+            items.push(CollectionItem::Collection(LayerCollectionListing {
+                id: ProviderLayerCollectionId {
+                    provider_id: NETCDF_CF_PROVIDER_ID,
+                    collection_id: NetCdfLayerCollectionId::Path {
+                        path: entry
+                            .path()
+                            .strip_prefix(base)
+                            .map_err(|_| Error::InvalidLayerCollectionId)?
+                            .to_owned(),
+                    }
+                    .try_into()?,
+                },
+                name: entry.file_name().to_string_lossy().to_string(),
+                description: "".to_string(),
+                entry_label: None,
+                properties: None,
+            }));
+        }
+    }
+
+    Ok(items)
+}
+
+/// find the group given by the path `groups`.
+pub fn find_group(
+    netcdf_groups: Vec<NetCdfGroup>,
+    groups: &[String],
+) -> crate::error::Result<Option<NetCdfGroup>> {
+    if groups.is_empty() {
+        return Ok(None);
+    }
+
+    let mut group_stack = groups.iter().collect::<Vec<_>>();
+    let mut group = netcdf_groups
+        .into_iter()
+        .find(|g| g.name == *group_stack.remove(0))
```

Comment:
If you always do remove(0), shouldn't you just once reverse the vector and then do pop?

Reply:
because it is less code this way

Reply:
Then just use reverse once and pop instead of remove(0)?

Reply:
that would be more code and I doubt it makes a difference performance-wise on a vector of length 3, but I can do it if you insist.

Reply:
Okay, keep it.

```Rust
use std::collections::VecDeque;

use criterion::{black_box, Criterion};
use criterion::{criterion_group, criterion_main};

pub fn vec_benchmark(c: &mut Criterion) {
    fn bench(values: [String; 3]) -> Vec<String> {
        let mut v: Vec<String> = Vec::from(values);
        while !v.is_empty() {
            v.remove(0);
        }
        v
    }

    c.bench_function("vec", |b| {
        b.iter(|| {
            bench(black_box([
                "foo".to_string(),
                "bar".to_string(),
                "baz".to_string(),
            ]))
        })
    });
}

pub fn vecdeque_benchmark(c: &mut Criterion) {
    fn bench(values: [String; 3]) -> VecDeque<String> {
        let mut v: VecDeque<String> = VecDeque::from(values);
        while !v.is_empty() {
            v.pop_front();
        }
        v
    }

    c.bench_function("vecdeque", |b| {
        b.iter(|| {
            bench(black_box([
                "foo".to_string(),
                "bar".to_string(),
                "baz".to_string(),
            ]))
        })
    });
}

pub fn vec_rev_benchmark(c: &mut Criterion) {
    fn bench(values: [String; 3]) -> Vec<String> {
        let mut v: Vec<String> = Vec::from(values);
        v.reverse();
        while !v.is_empty() {
            v.pop();
        }
        v
    }

    c.bench_function("vec_reverse", |b| {
        b.iter(|| {
            bench(black_box([
                "foo".to_string(),
                "bar".to_string(),
                "baz".to_string(),
            ]))
        })
    });
}

criterion_group!(
    benches,
    vec_benchmark,
    vecdeque_benchmark,
    vec_rev_benchmark
);
criterion_main!(benches);

```

```
Gnuplot not found, using plotters backend
vec                     time:   [32.511 ns 32.568 ns 32.642 ns]                 
Found 4 outliers among 100 measurements (4.00%)
  2 (2.00%) high mild
  2 (2.00%) high severe

vecdeque                time:   [38.337 ns 38.381 ns 38.425 ns]                      
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) low mild

vec_reverse             time:   [31.618 ns 31.656 ns 31.697 ns]                         
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) high mild
  1 (1.00%) high severe
```